Source/DafnyCore/ConcreteSyntax/ConcreteSyntaxTree.cs

```diff
@@ -73,6 +73,31 @@ public T Append<T>(T node)
       return this;
     }
 
+    public ConcreteSyntaxTree Comma<T>(IEnumerable<T> elements, Action<T> a) {
+      return Comma(", ", elements, (element, _) => a(element));
+    }
+
+    public ConcreteSyntaxTree Comma<T>(IEnumerable<T> elements, Action<T, int> a) {
+      return Comma(", ", elements, a);
+    }
+
+    public ConcreteSyntaxTree Comma<T>(string comma, IEnumerable<T> elements, Action<T> a) {
+      return Comma(comma, elements, (element, _) => a(element));
+    }
+
+    public ConcreteSyntaxTree Comma<T>(string comma, IEnumerable<T> elements, Action<T, int> a) {
+      var sep = "";
+      var index = 0;
+      foreach (var element in elements) {
+        Write(sep);
+        a(element, index);
+        sep = comma;
+        index++;
+      }
+
+      return this;
```

Comment:
These look useful. In fact, so useful that they might be used outside the compilers. I suggest moving the `Action`-based `Comma` methods to class `DafnyCore/Generic/Util.cs`, where there are already several other `Comma` methods. It would then make sense to make these new routines return a `string`, which the `ConcreteSyntaxTree` clients can then `Write(...)` themselves.

Reply:
No these versions are useful precisely it makes multiple writes to a `ConcreteSyntaxTree` instead of building up a string. It's useful for cases where the body of the "loop" is much more efficiently or naturally expressed by writing to a   `ConcreteSyntaxTree` (often by calling a fairly involved method like `TrExpr` which will write multiple times) instead of producing a string value. In cases where the body can return a string instead, then as you say you can just do something like `wr.Write(Comma(things, thing => MakeAString(thing)))` instead with the existing `Util.Comma` methods.

In a more recent commit I refactored some existing code to use it, which I think is a good illustrative example: https://github.com/dafny-lang/dafny/pull/3630/files#diff-703a5c7aa014afff3f25a9b9ebaee0546db822636c524b905406db519b2c1663L4709-R4729